### PR TITLE
Introduce Encoder API

### PR DESCRIPTION
**Pull request by rprimet**

### CODE REVIEW

File: alphacsc/encoder.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        """
+        raise NotImplementedError()
+    
+    def set_reg(self, reg):
+        """
+        Update the regularization parameter
+        """
+        raise NotImplementedError()
+    
+    def add_one_atom_in_z(self):
```
Comment (tomMoral) [inline alphacsc/encoder.py:83]: Maybe just use `add_one_atom(self, d_k)` that should add an atom in `D_hat` and add a 0 line in `z_hat`

File: alphacsc/learn_d_z_multi.py [hunk 2]
```diff
@@ -316,29 +297,35 @@ def _batch_learn(X, D_hat, z_hat, compute_z_func, compute_d_func,
...
             # add a new atom every n_iter_by_atom iterations
-            new_atom = get_max_error_dict(X, z_hat, D_hat,
+            new_atom = get_max_error_dict(X, z_encoder.get_z_hat(), D_hat,
                                           uv_constraint=uv_constraint,
                                           window=window)[0]
             D_hat = np.concatenate([D_hat, new_atom[None]])
-            z_hat = lil.add_one_atom_in_z(z_hat)
+            z_encoder.set_D(D_hat)
+            # XXX what should happen here when using DiCoDiLe?
+            z_encoder.add_one_atom_in_z()
```
Comment (tomMoral) [inline alphacsc/learn_d_z_multi.py:306]: Here, you should use `add_one_atom(new_atom)`

File: alphacsc/learn_d_z_multi.py [hunk 3]
```diff
@@ -217,66 +200,68 @@ def compute_d_func(X, z_hat, D_hat, constants):
...
 
-    if callable(callback):
-        callback(X, D_hat, z_hat, [])
+    with get_z_encoder_for(solver_z, z_kwargs, X, D_hat, n_atoms, n_times_atom, algorithm, reg, loss, loss_params, uv_constraint, feasible_evaluation=True, n_jobs=n_jobs) as z_encoder:
+        if callable(callback): 
+            callback(X, D_hat, z_encoder.get_z_hat(), [])
 
-    end_iter_func = get_iteration_func(eps, stopping_pobj, callback, lmbd_max,
+        end_iter_func = get_iteration_func(eps, stopping_pobj, callback, lmbd_max,
                                        name, verbose, raise_on_increase)
```
Comment (tomMoral) [inline alphacsc/learn_d_z_multi.py:208]: pep8
```suggestion
                                           name, verbose, raise_on_increase)
```

**Comment by tomMoral** [comment]
Things that we can do/discuss for `alphacsc/dicodile` integration. Let's discuss those on monday.

- Overall, avoid calling `get_z_hat` if not strictly necessary. This is costly in communication and all operations that can be distributed are more efficient.
- In `compute_d_multi`, give `z_encoder` instead of `X, z_hat, D_hat, constants`. This way, depending on the algorithm, it can either call `get_sufficient_statistic` or `get_z_hat` depending on what it needs. Maybe also we can see if we can implement some other operations in `update_d_multi` this way.
- For callback, also try to use the `z_encoder` as much as possible.
- implement `add_one_atom` in `dicodile`
- implement `get_error_max` in `dicodile` in a distributed algorithm
- implement efficient operation when `D_hat` is composed of rank_1  atoms in `dicodile`
- implement a multi signal dicodile version for online learning.

**Comment by rprimet** [comment]
Other remarks:

- docstrings for the encoders
- move encoder.py to _encoder.py (private API)

**Comment by rprimet** [comment]
Other remarks:

- docstrings for the encoders
- move encoder.py to _encoder.py (private API)

File: alphacsc/encoder.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
+import numpy as np
```
Comment (tomMoral) [inline alphacsc/encoder.py:1]: Rename the file `_encoder.py` to make it clear it is private API.

**Comment by rprimet** [comment]
Ensure `unbiased_z_hat` is handled (Encoder API) and add a test

**Comment by rprimet** [comment]
Consider adding `get_sufficient_statistics_partial` to the Encoder API

File: alphacsc/encoder.py (hunk 4 above)
Comment (rprimet) [reply alphacsc/encoder.py]: OK, done

File: alphacsc/_encoder.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        """
+        Perform one incremental z update.
+        This is the "main" function of the algorithm.
+        """
+        raise NotImplementedError()
+
+    def compute_z_partial(self, i0):
+        """
+        Compute z on a slice of the signal X,
+        for online learning.
```
Comment (tomMoral) [inline alphacsc/_encoder.py:131]: Doesn't this work on one line?
```suggestion
        Compute z on a slice of the signal X, for online learning.
```

File: alphacsc/_encoder.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+        """
+        raise NotImplementedError()
+
+    def get_sufficient_statistics(self):
+        """
+        Computes sufficient statistics to update D.
+
+        Returns
+        -------
+        ztz, ztX : (float, float)
```
Comment (tomMoral) [inline alphacsc/_encoder.py:156]: These are two ndarrays

File: alphacsc/_encoder.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+
+    def get_sufficient_statistics_partial(self):
+        """
+        Returns the partial sufficient statistics
+        that were computed during the last call to
+        compute_z_partial.
+
+        Returns
+        -------
+        ztz, ztX : (float, float)
```
Comment (tomMoral) [inline alphacsc/_encoder.py:169]: ditto

File: alphacsc/_encoder.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+        """
+        raise NotImplementedError()
+
+    def set_D(self, d):
+        """
+        Update the dictionary.
+
+        Parameters
+        ----------
+        d
```
Comment (tomMoral) [inline alphacsc/_encoder.py:192]: Maybe better to call this `D` ?
```suggestion
    def set_D(self, D):
        """
        Update the dictionary.

        Parameters
        ----------
        D : ndarray, shape (n_atoms, n_channels, n_time_atoms)
```

File: alphacsc/_encoder.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+        return self.ztz_i0, self.ztX_i0
+
+    def set_D(self, D):
+        self.D_hat = D
+
+    def set_reg(self, reg):
+        self.reg = reg
+
+    def add_one_atom(self, new_atom):
+        self.D_hat = np.concatenate([self.D_hat, new_atom[None]])
```
Comment (tomMoral) [inline alphacsc/_encoder.py:332]: Maybe add a small check for the shape of the added atom?

**Review by tomMoral — COMMENTED** [review]
Some comments on the PR, some can be taken into account in a follow up PR :)

File: alphacsc/_encoder.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    Parameters
+    ----------
+    solver : str
+        The solver to use for the z update. Options are
+        {{'l_bfgs' | 'lgcd'}}.
+    z_kwargs : dict
+        Additional keyword arguments to pass to update_z_multi.
+    X : array, shape (n_trials, n_channels, n_times)
+        The data on which to perform CSC.
+    D_hat : array, shape e(n_trials, n_channels, n_times) or
```
Comment (agramfort) [inline alphacsc/_encoder.py:35]:
```suggestion
    D_hat : array, shape (n_trials, n_channels, n_times) or
```
also I doubt the line below renders fine with numpydoc / sphinx

File: alphacsc/_encoder.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        (n_atoms, n_channels + atom_support)
+        The dictionary used to encode the signal X. Can be either in the form
+        f a full rank dictionary D (n_atoms, n_channels, atom_support) or with
```
Comment (agramfort) [inline alphacsc/_encoder.py:38]:
```suggestion
        of a full rank dictionary D (n_atoms, n_channels, atom_support) or with
```

File: alphacsc/_encoder.py [hunk 12]
```diff
@@ -0,0 +1,336 @@
...
+        i.e. norm(uv_hat) <= 1.
+    n_jobs : int
+        The number of parallel jobs.
+    use_sparse_z : boolean
+        Use sparse lil_matrices to store the activations.
+
+    Returns
+    -------
+    enc - a ZEncoder instance
+
```
Comment (agramfort) [inline alphacsc/_encoder.py:68]:
```suggestion
    enc : instance of ZEncoder
        The encoder.

```

we should run numpydoc checks in the repo.

File: alphacsc/_encoder.py [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+    Example usage
+    -------------
```
Comment (agramfort) [inline alphacsc/_encoder.py:70]: numpydoc says such a section should be called "Examples"

File: alphacsc/_encoder.py [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+
+    def compute_z_partial(self, i0):
+        """
+        Compute z on a slice of the signal X,
+        for online learning.
+
+        Parameters
+        ----------
+        i0 : int
+             Slice index.
```
Comment (agramfort) [inline alphacsc/_encoder.py:136]:
```suggestion
            Slice index.
```

File: alphacsc/_encoder.py (hunk 12 above)
Comment (rprimet) [reply alphacsc/_encoder.py:68]: OK, I'll add an issue for that

**Comment by agramfort** [comment]
thanks !

>

File: alphacsc/_encoder.py (hunk 13 above)
Comment (rprimet) [reply alphacsc/_encoder.py:71]: OK, I will remove the example altogether as it is unhelpful.

File: alphacsc/tests/test_encoder.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+N_TRIALS, N_CHANNELS, N_TIMES = 2, 3, 30
+N_TIMES_ATOM, N_ATOMS = 6, 4
+REG = 0.1
+
+rng = check_random_state(42)
+
+
+@pytest.fixture
+def X():
+    return rng.randn(N_TRIALS, N_CHANNELS, N_TIMES)
```
[the remark this answers is not in the document]
Comment (rprimet) [reply alphacsc/tests/test_encoder.py]: OK, made it a global for now.

File: alphacsc/tests/test_encoder.py [hunk 16]
```diff
@@ -0,0 +1,544 @@
...
+        get_z_encoder_for(solver=solver_z,
+                          z_kwargs=dict(),
+                          X=X,
+                          D_hat=D_hat,
+                          n_atoms=N_ATOMS,
+                          atom_support=N_TIMES_ATOM,
+                          algorithm='batch',
+                          reg=REG,
+                          loss='l2',
+                          loss_params=loss_params,
```
[the remark this answers is not in the document]
Comment (rprimet) [reply alphacsc/tests/test_encoder.py:71]: I see your point, and indeed having no default arguments in `get_z_encoder_for` makes it cumbersome. But setting defaults in `get_z_encoder_for` might be confusing too -- as the encoder is an internal class and doesn't really make assumptions, maybe the default arguments are better left to the level above, i.e. `learn_d_z_multi`? What do you think?

**Review by agramfort — APPROVED** [review]
LGTM

no objection to merge on my side if CIs are green

File: alphacsc/learn_d_z_multi.py [hunk 17]
```diff
@@ -200,11 +200,10 @@ def compute_d_func(X, z_hat, D_hat, constants):
                             uv_constraint=uv_constraint, loss=loss,
                             loss_params=loss_params, window=window, **d_kwargs)
 
-    with get_z_encoder_for(solver_z, z_kwargs, X, D_hat, n_atoms,
-                           n_times_atom, algorithm, reg, loss,
+    with get_z_encoder_for(X, D_hat, n_atoms, n_times_atom, n_jobs,
+                           solver_z, z_kwargs, algorithm, reg, loss,
                            loss_params, uv_constraint,
```
Comment (tomMoral) [inline alphacsc/learn_d_z_multi.py]: Use keyword argument whenever possible so there is no issue with argument ordering?

Comment (tomMoral) [reply alphacsc/learn_d_z_multi.py]: Maybe to change in another PR :)

**Review by tomMoral — APPROVED** [review]
LGTM! thx a lot for all the edits!
Let's merge this one and move on to the tasks from #42

**Comment by tomMoral** [comment]
Merging as the doc build should be solved in #44

**Pull request merged**